File: experiments/problem1_nonperiodicity/analyze_period_two_complete_local_quotient.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <limits>
#include <queue>
#include <string>
#include <unordered_map>
#include <vector>

using u128 = unsigned __int128;

std::string decimal_u128(u128 value) {
    if (value == 0) return "0";
    std::string digits;
    while (value > 0) {
        digits.push_back(static_cast<char>('0' + value % 10));
        value /= 10;
    }
    std::reverse(digits.begin(), digits.end());
    return digits;
}

namespace {

constexpr int SPAN = 5;
constexpr int MODULUS = 1 << (2 * SPAN);
constexpr int HISTORY_WIDTH = SPAN - 1;
constexpr int CERTIFICATE_STEPS = 112;
constexpr std::array<char, 3> LETTERS{'t', 'u', 'p'};

uint64_t forward_t(uint64_t state) {
    return state ^ ((state << 1) | (state << 2));
}

uint64_t forward_generator(int letter, uint64_t state) {
    const uint64_t stepped = forward_t(state);
    if (letter == 0) return stepped;
    if (letter == 1) return stepped ^ 1;
    return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
}

int power_int(int base, int exponent) {
    int result = 1;
    while (exponent-- > 0) result *= base;
    return result;
}

std::vector<int> decode_word(int code, int length) {
    std::vector<int> digits(length);
    for (int index = length - 1; index >= 0; --index) {
        digits[index] = code % 3;
        code /= 3;
    }
    return digits;
}

int apply_word_mod(int residue, int code, int length, int modulus) {
    for (int letter : decode_word(code, length)) {
        residue = static_cast<int>(forward_generator(letter, residue)) & (modulus - 1);
    }
    return residue;
}

struct LocalTables {
    std::array<std::vector<uint8_t>, SPAN + 1> minimal;
    std::array<int, SPAN + 1> word_counts{};
    std::array<int, SPAN + 1> residue_counts{};

    LocalTables() {
        for (int length = 1; length <= SPAN; ++length) {
            const int words = power_int(3, length);
            const int residues = 1 << (2 * length);
            word_counts[length] = words;
            residue_counts[length] = residues;
            minimal[length].assign(residues * words, 0);
            std::vector<int> best(residues, -1);
            for (int start = 0; start < residues; ++start) {
                std::fill(best.begin(), best.end(), -1);
                for (int word = 0; word < words; ++word) {
                    const int output = apply_word_mod(start, word, length, residues);
                    if (best[output] < 0) best[output] = word;
                }
                for (int word = 0; word < words; ++word) {
                    const int output = apply_word_mod(start, word, length, residues);
                    minimal[length][start * words + word] = (best[output] == word);
                }
            }
        }
    }

    bool is_minimal(int length, int start, int word) const {
        const int residue_mask = residue_counts[length] - 1;
        return minimal[length][(start & residue_mask) * word_counts[length] + word] != 0;
    }
};

struct State {
    uint16_t residue;
    uint8_t history_length;
    uint8_t history_code;
};

uint32_t state_key(const State& state) {
    return static_cast<uint32_t>(state.residue)
        | (static_cast<uint32_t>(state.history_length) << 10)
        | (static_cast<uint32_t>(state.history_code) << 13);
}

struct Automaton {
    LocalTables tables;
    std::array<std::array<uint16_t, MODULUS>, 3> inverse{};
    std::vector<State> states;
    std::vector<std::array<int32_t, 3>> transitions;
    std::unordered_map<uint32_t, int32_t> index;
    int32_t phase_p_start = -1;
    int32_t phase_u_start = -1;

    Automaton() {
        build_inverse();
        build_graph();
    }

    void build_inverse() {
        for (int letter = 0; letter < 3; ++letter) {
            std::array<bool, MODULUS> seen{};
            for (int state = 0; state < MODULUS; ++state) {
                const int image = static_cast<int>(forward_generator(letter, state)) & (MODULUS - 1);
                if (seen[image]) throw std::runtime_error("generator is not invertible");
                seen[image] = true;
                inverse[letter][image] = static_cast<uint16_t>(state);
            }
        }
    }

    int invert_suffix(int residue, int code, int length) const {
        const auto digits = decode_word(code, length);
        for (int index = length - 1; index >= 0; --index) {
            residue = inverse[digits[index]][residue];
        }
        return residue;
    }

    bool step(const State& state, int letter, State& output) const {
        const int new_residue = static_cast<int>(forward_generator(letter, state.residue)) & (MODULUS - 1);
        const int full_length = std::min<int>(SPAN, state.history_length + 1);
        const int full_code = state.history_code * 3 + letter;

        for (int length = 1; length <= full_length; ++length) {
            const int words = tables.word_counts[length];
            const int suffix = full_code % words;
            const int start = invert_suffix(new_residue, suffix, length);
            if (!tables.is_minimal(length, start, suffix)) return false;
        }

        output.residue = static_cast<uint16_t>(new_residue);
        output.history_length = static_cast<uint8_t>(std::min<int>(HISTORY_WIDTH, state.history_length + 1));
        output.history_code = static_cast<uint8_t>(full_code % power_int(3, HISTORY_WIDTH));
        return true;
    }

    int32_t add_state(const State& state, std::queue<int32_t>& queue) {
        const uint32_t key = state_key(state);
        const auto found = index.find(key);
        if (found != index.end()) return found->second;
        const int32_t id = static_cast<int32_t>(states.size());
        index.emplace(key, id);
        states.push_back(state);
        transitions.push_back({-2, -2, -2});
        queue.push(id);
        return id;
    }

    void build_graph() {
        index.reserve(30000);
        std::queue<int32_t> queue;
        phase_p_start = add_state(State{3, 1, 2}, queue);
        phase_u_start = add_state(State{1, 1, 1}, queue);
        while (!queue.empty()) {
            const int32_t source = queue.front();
            queue.pop();
            const State state = states[source];
            for (int letter = 0; letter < 3; ++letter) {
                State image{};
                transitions[source][letter] = step(state, letter, image)
                    ? add_state(image, queue)
                    : -1;
            }
        }
    }

    long double spectral_radius(int iterations = 500) const {
        std::vector<long double> vector(states.size(), 1.0L);
        std::vector<long double> updated(states.size());
        for (int iteration = 0; iteration < iterations; ++iteration) {
            long double scale = 0.0L;
            for (std::size_t source = 0; source < states.size(); ++source) {
                long double total = 0.0L;
                for (int32_t target : transitions[source]) {
                    if (target >= 0) total += vector[target];
                }
                updated[source] = total;
                scale = std::max(scale, total);
            }
            for (std::size_t position = 0; position < vector.size(); ++position) {
                vector[position] = updated[position] / scale;
            }
        }
        long double lower = std::numeric_limits<long double>::infinity();
        long double upper = 0.0L;
        for (std::size_t source = 0; source < states.size(); ++source) {
            long double total = 0.0L;
            for (int32_t target : transitions[source]) {
                if (target >= 0) total += vector[target];
            }
            if (vector[source] > 1e-24L) {
                const long double ratio = total / vector[source];
                lower = std::min(lower, ratio);
                upper = std::max(upper, ratio);
            }
        }
        return (lower + upper) / 2;
    }

    u128 maximum_continuations(int steps) const {
        std::vector<u128> current(states.size(), 1);
        std::vector<u128> updated(states.size());
        for (int step_index = 0; step_index < steps; ++step_index) {
            for (std::size_t source = 0; source < states.size(); ++source) {
                u128 total = 0;
                for (int32_t target : transitions[source]) {
                    if (target >= 0) total += current[target];
                }
                updated[source] = total;
            }
            current.swap(updated);
        }
        return *std::max_element(current.begin(), current.end());
    }

    uint64_t phase_count(int32_t start, int total_length) const {
        std::vector<uint64_t> current(states.size(), 0);
        std::vector<uint64_t> updated(states.size(), 0);
        current[start] = 1;
        for (int length = 2; length <= total_length; ++length) {
            std::fill(updated.begin(), updated.end(), 0);
            for (std::size_t source = 0; source < states.size(); ++source) {
                if (current[source] == 0) continue;
                for (int32_t target : transitions[source]) {
                    if (target >= 0) updated[target] += current[source];
                }
            }
            current.swap(updated);
        }
        uint64_t total = 0;
        for (uint64_t count : current) total += count;
        return total;
    }
};

}  // namespace

int main() {
    const Automaton automaton;
    const long double radius = automaton.spectral_radius();
    const u128 maximum = automaton.maximum_continuations(CERTIFICATE_STEPS);
    const u128 binary = u128{1} << CERTIFICATE_STEPS;
    if (!(maximum < binary)) {
        std::cerr << "below-binary certificate failed\n";
        return 2;
    }
    if (automaton.states.size() != 21615) {
        std::cerr << "unexpected reachable state count\n";
        return 3;
    }
    const u128 expected = (u128{0xff9da21103f2ULL} << 64) | u128{0xdca8bea29790526dULL};
    if (maximum != expected) {
        std::cerr << "unexpected continuation certificate\n";
        return 4;
    }

    std::size_t transition_count = 0;
    for (const auto& row : automaton.transitions) {
        transition_count += std::count_if(
            row.begin(), row.end(), [](int32_t value) { return value >= 0; });
    }

    std::cout << "span " << SPAN << '\n';
    std::cout << "modulus " << MODULUS << '\n';
    std::cout << "reachable_states " << automaton.states.size() << '\n';
    std::cout << "allowed_transitions " << transition_count << '\n';
    std::cout << std::setprecision(18);
    std::cout << "spectral_radius_approx " << static_cast<double>(radius) << '\n';
    std::cout << "generic_complexity_rate_approx "
              << static_cast<double>(std::log(2.0L) / std::log(radius)) << '\n';
    std::cout << "certificate_steps " << CERTIFICATE_STEPS << '\n';
    std::cout << "maximum_continuations " << decimal_u128(maximum) << '\n';
    std::cout << "binary_continuations " << decimal_u128(binary) << '\n';
    std::cout << "strictly_below_binary true\n";
    std::cout << "phase_p_length18 " << automaton.phase_count(automaton.phase_p_start, 18) << '\n';
    std::cout << "phase_u_length18 " << automaton.phase_count(automaton.phase_u_start, 18) << '\n';
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_complete_local_quotient.py
#!/usr/bin/env python3
"""Build complete bounded-span canonical quotients for period-two witnesses.

For a common block length m, two generator words can differ only in the lowest
2m output bits, and those bits depend only on the lowest 2m input bits. Thus
all same-length arithmetic identities are decided by one finite table modulo
4**m.

For a configured span s, this analyzer replaces every block of length at most s
by the lexicographically least equal block at its exact starting residue. The
resulting irreducible words form a finite-state language. Its growth bounds the
number of distinct arithmetic witnesses.

The Python implementation is capped at span four. The companion C++ analyzer
constructs the span-five graph and emits the exact below-binary path-count
certificate. These generic quotients do not prove divergence on the one
zero-initialized Rule 30 fringe schedule.
"""
from __future__ import annotations

import argparse
import hashlib
import itertools
import json
import math
from collections import defaultdict, deque
from dataclasses import dataclass
from typing import Any, Iterable

LETTERS = ("t", "u", "p")
LETTER_INDEX = {letter: index for index, letter in enumerate(LETTERS)}
PHASE_START = {"p": 3, "u": 1}
DEFAULT_SPAN = 3
ABSOLUTE_SPAN = 4
DEFAULT_COUNT_LENGTH = 14
ABSOLUTE_COUNT_LENGTH = 18


def forward_t(state: int) -> int:
    if state < 0:
        raise ValueError("state must be nonnegative")
    return state ^ ((state << 1) | (state << 2))


def forward_generator(letter: str, state: int) -> int:
    stepped = forward_t(state)
    if letter == "t":
        return stepped
    if letter == "u":
        return stepped ^ 1
    if letter == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def apply_word(state: int, word: Iterable[str]) -> int:
    for letter in word:
        state = forward_generator(letter, state)
    return state


def word_index(word: str) -> int:
    value = 0
    for letter in word:
        value = 3 * value + LETTER_INDEX[letter]
    return value


def words_of_length(length: int) -> tuple[str, ...]:
    return tuple("".join(word) for word in itertools.product(LETTERS, repeat=length))


def complete_local_minima(span: int) -> dict[tuple[int, int, str], str]:
    """Return every lex-min same-length representative through ``span``."""
    if not 1 <= span <= ABSOLUTE_SPAN:
        raise ValueError("span outside Python campaign cap")
    full_modulus = 1 << (2 * span)
    table: dict[tuple[int, int, str], str] = {}
    for length in range(1, span + 1):
        modulus = 1 << (2 * length)
        words = words_of_length(length)
        residue_minima: list[dict[int, str]] = []
        for residue in range(modulus):
            minima: dict[int, str] = {}
            for word in words:
                output = apply_word(residue, word) & (modulus - 1)
                minima.setdefault(output, word)
            residue_minima.append(minima)
        for lifted_residue in range(full_modulus):
            residue = lifted_residue & (modulus - 1)
            minima = residue_minima[residue]
            for word in words:
                output = apply_word(residue, word) & (modulus - 1)
                table[(lifted_residue, length, word)] = minima[output]
    return table


def inverse_tables(modulus: int) -> dict[str, tuple[int, ...]]:
    result: dict[str, tuple[int, ...]] = {}
    for letter in LETTERS:
        inverse = [-1] * modulus
        for state in range(modulus):
            image = forward_generator(letter, state) & (modulus - 1)
            if inverse[image] != -1:
                raise AssertionError("generator is not invertible modulo power of two")
            inverse[image] = state
        if any(value < 0 for value in inverse):
            raise AssertionError("generator permutation is incomplete")
        result[letter] = tuple(inverse)
    return result


def invert_word(residue: int, word: str, inverse: dict[str, tuple[int, ...]]) -> int:
    for letter in reversed(word):
        residue = inverse[letter][residue]
    return residue


@dataclass(frozen=True, order=True)
class AutomatonState:
    residue: int
    history: str


class CompleteLocalAutomaton:
    def __init__(self, span: int):
        if not 1 <= span <= ABSOLUTE_SPAN:
            raise ValueError("span outside Python campaign cap")
        self.span = span
        self.modulus = 1 << (2 * span)
        self.mask = self.modulus - 1
        self.minima = complete_local_minima(span)
        self.inverse = inverse_tables(self.modulus)
        self.phase_starts = {
            phase: AutomatonState(state & self.mask, phase)
            for phase, state in PHASE_START.items()
        }
        self.states, self.transitions = self._build_reachable_graph()
        self.index = {state: index for index, state in enumerate(self.states)}

    def step(self, state: AutomatonState, letter: str) -> AutomatonState | None:
        if letter not in LETTER_INDEX:
            raise ValueError("unknown generator")
        new_residue = forward_generator(letter, state.residue) & self.mask
        full_history = state.history + letter
        for length in range(1, min(self.span, len(full_history)) + 1):
            block = full_history[-length:]
            start = invert_word(new_residue, block, self.inverse)
            if self.minima[(start, length, block)] != block:
                return None
        history_width = max(0, self.span - 1)
        new_history = full_history[-history_width:] if history_width else ""
        return AutomatonState(new_residue, new_history)

    def _build_reachable_graph(
        self,
    ) -> tuple[tuple[AutomatonState, ...], dict[AutomatonState, tuple[AutomatonState | None, ...]]]:
        reached = set(self.phase_starts.values())
        queue: deque[AutomatonState] = deque(reached)
        transitions: dict[AutomatonState, tuple[AutomatonState | None, ...]] = {}
        while queue:
            state = queue.popleft()
            images = tuple(self.step(state, letter) for letter in LETTERS)
            transitions[state] = images
            for image in images:
                if image is not None and image not in reached:
                    reached.add(image)
                    queue.append(image)
        states = tuple(sorted(reached))
        for state in states:
            transitions.setdefault(state, tuple(self.step(state, letter) for letter in LETTERS))
        return states, transitions

    def accepted(self, phase: str, suffix: str) -> bool:
        state = self.phase_starts[phase]
        for letter in suffix:
            image = self.step(state, letter)
            if image is None:
                return False
            state = image
        return True

    def count_by_phase(self, maximum_length: int) -> list[dict[str, int]]:
        if maximum_length < 1:
            raise ValueError("maximum length must be positive")
        current = {phase: {start: 1} for phase, start in self.phase_starts.items()}
        rows: list[dict[str, int]] = []
        for length in range(1, maximum_length + 1):
            if length > 1:
                for phase in current:
                    updated: defaultdict[AutomatonState, int] = defaultdict(int)
                    for state, count in current[phase].items():
                        for image in self.transitions[state]:
                            if image is not None:
                                updated[image] += count
                    current[phase] = dict(updated)
            rows.append({
                "length": length,
                **{phase: sum(row.values()) for phase, row in current.items()},
            })
        return rows

    def spectral_radius(self, iterations: int = 300) -> float:
        vector = [1.0] * len(self.states)
        for _ in range(iterations):
            updated = [0.0] * len(self.states)
            for source, state in enumerate(self.states):
                updated[source] = sum(
                    vector[self.index[image]]
                    for image in self.transitions[state]
                    if image is not None
                )
            scale = max(updated)
            if scale == 0:
                return 0.0
            vector = [value / scale for value in updated]
        ratios = []
        for source, state in enumerate(self.states):
            numerator = sum(
                vector[self.index[image]]
                for image in self.transitions[state]
                if image is not None
            )
            if vector[source] > 1e-15:
                ratios.append(numerator / vector[source])
        return (min(ratios) + max(ratios)) / 2


def reduce_word(word: str, span: int, minima: dict[tuple[int, int, str], str]) -> tuple[str, int]:
    """Reduce one normalized word by the first available local replacement."""
    if not word or word[0] not in PHASE_START:
        raise ValueError("word must be normalized and begin with p or u")
    full_mask = (1 << (2 * span)) - 1
    steps = 0
    while True:
        state = 0
        changed = False
        for index in range(len(word)):
            for length in range(1, min(span, len(word) - index) + 1):
                block = word[index : index + length]
                minimum = minima[(state & full_mask, length, block)]
                if minimum != block:
                    candidate = word[:index] + minimum + word[index + length :]
                    if word_index(candidate) >= word_index(word):
                        raise AssertionError("local replacement did not lower the word")
                    if apply_word(0, candidate) != apply_word(0, word):
                        raise AssertionError("local replacement changed arithmetic state")
                    word = candidate
                    steps += 1
                    changed = True
                    break
            if changed:
                break
            state = forward_generator(word[index], state)
        if not changed:
            return word, steps


def verify_locality_tables(span: int) -> dict[str, Any]:
    minima = complete_local_minima(span)
    checks = 0
    nontrivial = 0
    for length in range(1, span + 1):
        modulus = 1 << (2 * length)
        words = words_of_length(length)
        for residue in range(modulus):
            for word in words:
                minimum = minima[(residue, length, word)]
                if apply_word(residue, word) & (modulus - 1) != apply_word(
                    residue, minimum
                ) & (modulus - 1):
                    raise AssertionError("local minimum table is not output preserving")
                checks += 1
                nontrivial += int(minimum != word)
    return {
        "span": span,
        "finite_table_checks": checks,
        "nonminimal_entries": nontrivial,
        "locality_lemma": (
            "equal-length-m word differences are confined below bit 2m, so the "
            "table modulo 4^m is an all-width identity table"
        ),
        "all_checks_pass": True,
    }


def verify_reduction(span: int, maximum_length: int) -> dict[str, Any]:
    minima = complete_local_minima(span)
    automaton = CompleteLocalAutomaton(span)
    checked = changed = maximum_steps = 0
    for length in range(1, maximum_length + 1):
        for phase in PHASE_START:
            for suffix in itertools.product(LETTERS, repeat=length - 1):
                word = phase + "".join(suffix)
                reduced, steps = reduce_word(word, span, minima)
                if not automaton.accepted(reduced[0], reduced[1:]):
                    raise AssertionError("reduced word is not accepted")
                checked += 1
                changed += int(steps > 0)
                maximum_steps = max(maximum_steps, steps)
    return {
        "maximum_normalized_length": maximum_length,
        "words_checked": checked,
        "words_changed": changed,
        "maximum_rewrite_steps": maximum_steps,
        "all_checks_pass": True,
    }


def exact_arithmetic_counts(maximum_length: int) -> list[dict[str, int]]:
    current = {phase: {state} for phase, state in PHASE_START.items()}
    rows: list[dict[str, int]] = []
    for length in range(1, maximum_length + 1):
        if length > 1:
            for phase in current:
                current[phase] = {
                    forward_generator(letter, state)
                    for state in current[phase]
                    for letter in LETTERS
                }
        rows.append({
            "length": length,
            **{phase: len(states) for phase, states in current.items()},
        })
    return rows


def run_campaign(*, span: int = DEFAULT_SPAN, count_length: int = DEFAULT_COUNT_LENGTH) -> dict[str, Any]:
    if not 1 <= span <= ABSOLUTE_SPAN:
        raise ValueError("span outside Python campaign cap")
    if not 1 <= count_length <= ABSOLUTE_COUNT_LENGTH:
        raise ValueError("count length outside campaign cap")
    automaton = CompleteLocalAutomaton(span)
    canonical_counts = automaton.count_by_phase(count_length)
    arithmetic_counts = exact_arithmetic_counts(count_length)
    for canonical, arithmetic in zip(canonical_counts, arithmetic_counts, strict=True):
        for phase in PHASE_START:
            if arithmetic[phase] > canonical[phase]:
                raise AssertionError("canonical language does not bound arithmetic image")
    radius = automaton.spectral_radius()
    payload: dict[str, Any] = {
        "span": span,
        "modulus": automaton.modulus,
        "reachable_automaton_states": len(automaton.states),
        "spectral_radius_approx": radius,
        "generic_complexity_rate_approx": math.log(2) / math.log(radius),
        "locality_tables": verify_locality_tables(span),
        "reduction_checks": verify_reduction(span, min(8, count_length)),
        "canonical_counts": canonical_counts,
        "arithmetic_counts": arithmetic_counts,
        "exact_conclusions": {
            "complete_local_reduction": (
                "every normalized word has a same-length representative whose every "
                "block through the configured span is lex-minimal at its exact residue"
            ),
            "regular_language_bound": (
                "the fixed-span irreducible language is finite-state and bounds the "
                "fixed-phase arithmetic image"
            ),
            "remaining_target": (
                "the quotient is generic and does not exclude the deterministic "
                "zero-initialized fringe survivor"
            ),
        },
        "scope_warning": (
            "the Python campaign is capped at span four; the companion C++ campaign "
            "provides the span-five below-binary certificate"
        ),
    }
    digest = hashlib.sha256()
    digest.update(b"rule30-period-two-complete-local-quotient-v1\0")
    digest.update(json.dumps(payload, sort_keys=True, separators=(",", ":")).encode())
    payload["certificate_sha256"] = digest.hexdigest()
    return payload


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--span", type=int, default=DEFAULT_SPAN)
    parser.add_argument("--count-length", type=int, default=DEFAULT_COUNT_LENGTH)
    args = parser.parse_args()
    print(json.dumps(run_campaign(span=args.span, count_length=args.count_length), sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
